### PR TITLE
Decimal Writer Support

### PR DESCRIPTION
**Pull request by wilwade**
Problem
=======
Need to support writing decimal types

Also closes: #87 

Solution
========

Add basic encoding support for decimals

Change summary:
---------------
* Added better decimal field errors
* Default scale to 0 per spec
* Cleanup types on RLE so it only asks for what is needed
* Support decimal encoding
* Add test for write / read of decimal

Steps to Verify:
----------------
1. Generate a schema with decimal field
2. Use it!

### CODE REVIEW

File: lib/codec/rle.ts
Comment (wilwade) [inline lib/codec/rle.ts:1]: This is just setting the interface to only what the things need. Much easier for fixing other types later

File: lib/codec/plain_dictionary.ts [hunk 1]
```diff
@@ -2,7 +2,7 @@ import * as rle from './rle'
 import { Cursor, Options } from './types'
 
 export const decodeValues = function(type: string, cursor: Cursor, count: number, opts: Options) {
-  opts.bitWidth = cursor.buffer.slice(cursor.offset, cursor.offset+1).readInt8(0);
+  const bitWidth = cursor.buffer.slice(cursor.offset, cursor.offset+1).readInt8(0);
   cursor.offset += 1;
-  return rle.decodeValues(type, cursor, count, Object.assign({}, opts, {disableEnvelope: true}));
+  return rle.decodeValues(type, cursor, count, Object.assign({}, opts, { disableEnvelope: true, bitWidth }));
```
Comment (wilwade) [inline lib/codec/plain_dictionary.ts:7]: Better than modifying the opts, and now types are happier.

File: lib/codec/types.ts [hunk 2]
```diff
@@ -5,7 +5,7 @@ import { Statistics } from "../../gen-nodejs/parquet_types";
 export interface Options {
     typeLength: number,
     bitWidth: number,
-    disableEnvelope: boolean
+    disableEnvelope?: boolean
```
Comment (wilwade) [inline lib/codec/types.ts:8]: This is always used as a default false, so ok to be optional.

File: lib/schema.ts
Comment (wilwade) [inline lib/schema.ts:1]: Better field errors. Also this is what fixes #87

File: lib/codec/rle.ts [hunk 3]
```diff
@@ -3,9 +3,9 @@
 // https://github.com/apache/parquet-format/blob/master/Encodings.md
```
Comment (wilwade) [inline lib/codec/rle.ts:3]: Updates to this file are just setting the interface to only what the things need. Much easier for fixing other types later

File: lib/schema.ts [hunk 4]
```diff
@@ -170,6 +170,8 @@ function buildFields(schema: SchemaDefinition, rLevelParentMax?: number, dLevelP
     }
 
     if (typeDef.originalType === 'DECIMAL') {
+      // Default scale to 0 per https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#decimal
+      if (typeof opts.scale === "undefined") opts.scale = 0;
```
Comment (wilwade) [inline lib/schema.ts:174]: Better field errors in this file and this is what fixes #87

**Review by enddynayn — APPROVED** [review]
fantastic! 👍

**Comment by JasonYeMSFT** [comment]
The changes look good regarding #87. Please feel free to close the issue once this PR is merged.

**Pull request merged**